engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/bool/OrNode.java

```diff
@@ -1,14 +1,41 @@
 package org.enso.interpreter.node.expression.builtin.bool;
 
 import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.profiles.ConditionProfile;
 import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.dsl.MonadicState;
+import org.enso.interpreter.dsl.Suspend;
+import org.enso.interpreter.node.BaseNode;
+import org.enso.interpreter.node.callable.thunk.ThunkExecutorNode;
+import org.enso.interpreter.runtime.Context;
+import org.enso.interpreter.runtime.error.DataflowError;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.enso.interpreter.runtime.state.Stateful;
 
 @BuiltinMethod(
     type = "Boolean",
     name = "||",
     description = "Computes the logical disjunction of two booleans")
 public class OrNode extends Node {
-  boolean execute(boolean _this, boolean that) {
-    return _this || that;
+  private @Child ThunkExecutorNode rhsThunkExecutorNode = ThunkExecutorNode.build();
+  private final ConditionProfile condProfile = ConditionProfile.createCountingProfile();
+
+  Stateful execute(@MonadicState Object state, boolean _this, @Suspend Object that) {
+    if (condProfile.profile(_this)) {
+      return new Stateful(state, true);
+    } else {
+      return ensureBooleanOrDataflowError(
+          rhsThunkExecutorNode.executeThunk(that, state, BaseNode.TailStatus.TAIL_DIRECT));
+    }
+  }
+
+  private Stateful ensureBooleanOrDataflowError(Stateful v) {
```

Comment:
Shouldn't this rather be split to three `@Specialization`s? Specialize on `Boolean`, specialize on `DataflowError`, `makeTypeError` as a fallback? Otherwise whole code of this method is going to be part of the compilation of each logical operator, I am afraid.

Reply:
Good point. Rewrote it